odxtools/ecu_variant_patterns.py

```diff
@@ -0,0 +1,84 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2023 MBition GmbH
+
+from dataclasses import dataclass
+from typing import List, Optional, Union
+from xml.etree import ElementTree
+
+from odxtools.odxlink import OdxDocFragment
+
+
+@dataclass
+class MatchingParameter:
+    """According to ISO 22901, a MatchingParameter contains a string value identifying
+    the active ECU variant. Moreover, it references a DIAG-COMM via snref and one of its
+    positiv response's OUT-PARAM-IF via snref or snpathref.
+
+    Unlike other parameters defined in the `parameters` package, a MatchingParameter is
+    not transferred over the network.
+    """
+
+    expected_value: str  # datatype according to ISO 22901-1 Figure 141
+    diag_comm_snref: str
+    out_param_if_snref: str
+
+    @classmethod
+    def from_et(
+        cls, et_element: ElementTree.Element, doc_frags: List[OdxDocFragment]
+    ) -> "MatchingParameter":
+
+        expected_value = et_element.findtext("EXPECTED-VALUE")
+        assert expected_value is not None
+
+        diag_com_snref_el = et_element.find("DIAG-COMM-SNREF")
+        assert diag_com_snref_el is not None
+        diag_comm_snref = diag_com_snref_el.get("SHORT-NAME")
+        assert diag_comm_snref is not None
+
+        out_param_snref_el = et_element.find("OUT-PARAM-IF-SNREF")
+        assert out_param_snref_el is not None
+        out_param_if_snref = out_param_snref_el.get("SHORT-NAME")
+        assert out_param_if_snref is not None
+
+        return cls(expected_value, diag_comm_snref, out_param_if_snref)
+
+    def is_match(self, ident_value: Union[int, str]) -> bool:
```

Comment:
Not possible for ident_value to be `bytes` ?

Reply:
Hmm, difficult... According to the ISO, the expected value is of type string. Thus, the `ident_value` should also be of type `str` in order to compare the values.

During the tests, I realized that odxtools convert numeric (integer) values from responses of a `DiagService` to `int`. As a workaround, I converted these values to `str`, which is straightforward. 

If the `ident_value` was of type `bytes`, we would have to know its encoding, which we do not know here. That does not imply, that the response's DOP cannot be a HexDump or something similar. We only need to make sure, that we can derive a string representation from whatever DOP it is. What Python datatype do odxtools use to represent a HexDump? @kayoub5 @andlaus
I guess we have similar issues with non-integer numeric values, such as float. Casting could be dangerous, since the decimal separator could be `.` or `,` in the string representation...

Having now thought more about this in greater detail, I guess it is the best if we strictly demand a `str` here. The conversion is deferred to the caller, who has likely more context about the DOP than we have here.

Reply:
float is not likely to be used since float equality is always a problematic topic.

for bytes we could assume an ascii encoding, not likely anyone will use non latin characters in an ecu identification

Reply:
> During the tests, I realized that odxtools convert numeric (integer) values from responses of a DiagService to int. As a workaround, I converted these values to str, which is straightforward.

I suppose that what it is basically depends on what your PDX defines as physical type. I guess it would be better either to define this argument as `Any` and demand that it can be compared, or pass the corresponding `Parameter` object along with it.

> What Python datatype do odxtools use to represent a HexDump?

Maybe I am missing something here (ODX is even more twisted than usual in that regard), but I think this works via the parameter's `dop`: conversion of the raw values from the wire to the internal type is done via the DOPs [`convert_physical_to_internal()`](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/dataobjectproperty.py#L167) method, the reverse is basically done using [`convert_physical_to_bytes()`](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/dataobjectproperty.py#L63) which might be what you want...